compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.cpp

```diff
@@ -0,0 +1,449 @@
+   *                         |
+   *   analyze_transpose     |    - value :   (1)   (448)  (7)  (7)
+   *                         |    - tags  :   (-)   (2)    (0)  (1)
+   *                         |
+   *   analyze_reshape       |    - value :   (1)   (448)   (49)
+   *                         |    - tags  :   (-)   (2)     (0, 1)
+   *                         |
+   *   anaylze_transpose     |    - value  :   (1)   (49)    (448)
+   *                         |    - tags   :   (-)   (0, 1)  (2)
+   *
+   *  After all simulation done, if tags are in same order as initial _shape,
+   *  Transpose has no effect in final shape, which they can be removed as
+   *  unnecessary Ops.
+   */
+  template <loco::DataType DType>
+  bool can_remove_transposes(const luci::CircleTranspose *f_tr, const luci::CircleReshape *m_rs,
+                             const luci::CircleTranspose *b_tr);
+
+private:
+  void init_shape_with_tag(const luci::CircleNode *in_tensor);
+
+  template <loco::DataType PermType>
+  bool analyze_transpose(const luci::CircleTranspose *transpose_node);
+
+  template <loco::DataType ShapeType> bool analyze_reshape(const luci::CircleReshape *reshape_node);
+
+  bool verify_tag() const;
+
+  struct Dim final
+  {
+    uint32_t value;
+    std::vector<uint8_t> tags;
+  };
+  using Shape = std::vector<Dim>;
+
+  const uint8_t START_TAG = 0;
+  Shape _shape;
+};
+
+/**
+ * @brief initalize _shape with input tensor named in_tensor
+ *
+ * @note 'tags' are attached to non-1 valued dimension.
+ */
+void TaggedShapeAnalyzer::init_shape_with_tag(const luci::CircleNode *in_tensor)
+{
+  _shape.clear();
+  uint8_t tag = START_TAG;
+
+  for (uint32_t i = 0; i < in_tensor->rank(); i++)
+  {
+    TaggedShapeAnalyzer::Dim dim;
+    {
+      dim.value = in_tensor->dim(i).value();
+      if (dim.value != 1)
+        dim.tags.push_back(tag++);
+    }
+    _shape.push_back(dim);
+  }
+}
+
+/**
+ * @brief update _shape based on 'Transpose' permutation value
+ *
+ * @example Let's assume Transpose(perm=0, 3, 1, 2) is given to [before] _shape.
+ *
+ *  This function reordered the Dims' order based on permutaiton value.
+ *
+ *  [before]  _shape :
+ *              - value :   (1)   (7)   (7)   (448)
+ *              - tags  :   (-)   (0)   (1)   (2)
+ *
+ *  [after]  _shape :
+ *              - value :   (1)   (448)  (7)  (7)
+ *              - tags  :   (-)   (2)    (0)  (1)
+ */
+template <loco::DataType PermType>
+bool TaggedShapeAnalyzer::analyze_transpose(const luci::CircleTranspose *transpose_node)
+{
+  const luci::CircleConst *perm_node = loco::must_cast<luci::CircleConst *>(transpose_node->perm());
+  assert(perm_node->dtype() == PermType);
+
+  TaggedShapeAnalyzer::Shape new_shape;
+  const auto size = perm_node->size<PermType>();
+  for (uint32_t i = 0; i < size; i++)
+  {
+    auto perm_idx = perm_node->at<PermType>(i);
+    new_shape.push_back(_shape.at(perm_idx));
+  }
+  _shape = new_shape;
+  return true;
```

Comment:
Does this method need to return boolean value?

Reply:
It always returns `true`, So yes.. It is meaningless. 
 
I intentionally make it return `boolean` to have a similar signature with `analyze_reshape` function.

Reply:
Do you think it is better to make it return `void`? 
About this, I don't have any strong opinion and I'd like to hear your preference. :smiley:

Reply:
I re-think this part, and i decided to make this function return `void`.

Thank you for review! :smile: